### PR TITLE
Add WithCancellation for async enumerables

### PR DESCRIPTION
**Pull request by stephentoub**
Contributes to https://github.com/dotnet/corefx/issues/33909
cc: @jcouv, @kouvel, @tarekgh 

Relates to async-streams umbrella: https://github.com/dotnet/roslyn/issues/24037

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/ConfiguredCancelableAsyncEnumerable.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>Provides an awaitable async enumerable that enables cancelable iteration and configured awaits.</summary>
+    [StructLayout(LayoutKind.Auto)]
+    public readonly struct ConfiguredCancelableAsyncEnumerable<T>
```
Comment (jcouv) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/ConfiguredCancelableAsyncEnumerable.cs]: >ConfiguredCancelableAsyncEnumerable [](start = 27, length = 35)

nit: I'm unsure which spelling is more common (cancellable vs. cancelable). I would have gone with two L's.

Comment (tarekgh) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/ConfiguredCancelableAsyncEnumerable.cs]: We have it one L in other places like TaskCanceledException, Task.IsCanceled, Task.FromCanceled, TaskCompletionSource.SetCanceled..etc. 

Although we are not consistent in the spelling as CancellationToken uses 2 L's but it looks the majority of the usage is using 1 L.

**Comment by tarekgh** [comment]
> I don't know how testing works. I assume you'll have some coverage in another PR/repo?

@jcouv you already included in the test changes PR https://github.com/dotnet/corefx/pull/34528

**Comment by onovotny** [comment]
This PR is a bit problematic when trying to copy these types into a downlevel platform and support unification with type forwarders. The problem is twofold:

1. The internal constructors. If I copy the type, I can't correctly type forward since the new type's ctor won't be visible. 
2. Adding methods to TaskExtensions. I can't forward certain methods; it would be much easier to add those to a new type that can be copied in full and typeforwarded on the netstandard2.1/netcoreapp3.0 version.

Can the ctor's be made public and the new methods from TaskExtension be split into its own type?

**Pull request merged**